tapdance/conn_raw.go

```diff
@@ -454,19 +456,22 @@ func (tdRaw *tdRawConn) readProto() (msg pb.StationToClient, err error) {
 	}
 
 	totalBytesToRead := headerSize + msgLen
-	readBuffer := make([]byte, msgLen)
+	var readBuffer bytes.Buffer
 
 	// Get the message itself
 	for readBytesTotal < totalBytesToRead {
-		readBytes, err = tdRaw.tlsConn.Read(readBuffer[readBytesTotal-headerSize : msgLen])
-		readBytesTotal += uint32(readBytes)
+		copiedBytes, err = io.CopyN(&readBuffer, tdRaw.tlsConn, int64(totalBytesToRead-readBytesTotal))
+		if copiedBytes > math.MaxUint32 {
+			panic("copiedBytes > MaxUint32")
```

Comment:
![image](https://user-images.githubusercontent.com/5443147/46897502-352a6600-ce40-11e8-999a-1f667ca72549.png)

Reply:
If I am reading https://golang.org/src/io/io.go?s=11939:12009#L329 correctly, we can just do `_, err = io.CopyN`, since it won't return without an error if it read less bytes than it was asked.